pkg/operator/reporting/validate.go

```diff
@@ -272,14 +324,29 @@ func NewReportDataSourceClientGetter(getter meteringClient.ReportDataSourcesGett
 	})
 }
 
-func GetDependentDataSources(dataSourceGetter reportDataSourceGetter, generationQuery *metering.ReportGenerationQuery) ([]*metering.ReportDataSource, error) {
-	dataSources := make([]*metering.ReportDataSource, len(generationQuery.Spec.DataSources))
-	for i, dataSourceName := range generationQuery.Spec.DataSources {
+func GetDependentDataSourcesMemoized(dataSourceGetter reportDataSourceGetter, generationQuery *metering.ReportGenerationQuery, dataSourceAccumulator map[string]*metering.ReportDataSource) error {
+	for _, dataSourceName := range generationQuery.Spec.DataSources {
+		if _, exists := dataSourceAccumulator[dataSourceName]; exists {
+			continue
+		}
 		dataSource, err := dataSourceGetter.getReportDataSource(generationQuery.Namespace, dataSourceName)
 		if err != nil {
-			return nil, err
+			return err
 		}
-		dataSources[i] = dataSource
+		dataSourceAccumulator[dataSource.Name] = dataSource
+	}
+	return nil
+}
+
+func GetDependentDataSources(dataSourceGetter reportDataSourceGetter, generationQuery *metering.ReportGenerationQuery) ([]*metering.ReportDataSource, error) {
+	dataSourceAccumulator := make(map[string]*metering.ReportDataSource)
+	err := GetDependentDataSourcesMemoized(dataSourceGetter, generationQuery, dataSourceAccumulator)
+	if err != nil {
+		return nil, err
+	}
+	dataSources := make([]*metering.ReportDataSource, 0, len(dataSourceAccumulator))
+	for _, ds := range dataSources {
```

Comment:
I'm a little confused by this for loop, but it may be me. Shouldn't we range over `dataSourceAccumulator` and put that in the `dataSources` slice?

Reply:
Yeah good catch.